### PR TITLE
HHH-16586 When merging a persisted entity with a null Version, Hibernate treats entity as transient instead of throwing an Exception

### PR DESCRIPTION
**Pull request by dreab8**
https://hibernate.atlassian.net/browse/HHH-16586

Not sure about the necessity of the `VersionValueException` (may be throwing an `HibernateException` is enough) neither if `VersionValueException`  is a good name and if the package I placed it is appropriate.

### CODE REVIEW

**Comment by gavinking** [comment]
Wait, wait wait, this is correct behavior, isn't it? An entity with a null version must be a new transient entity.

Are we fixing something here that we shouldn't fix?

File: hibernate-core/src/main/java/org/hibernate/exception/VersionValueException.java [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package org.hibernate.exception;
```
Comment (gavinking) [inline hibernate-core/src/main/java/org/hibernate/exception/VersionValueException.java:1]: So this is the wrong package for this exception. (Check the package-level javadoc.)

On the other hand, I'm not sure that this particular condition really deserves it's whole own exception type.

I would consider just using `PropertyValueException` for this; the semantics seem to match pretty closely.

Even if you think a dedicated exception type _is_ worthwhile, I would make `VersionValueException` extend `PropertyValueException`.

File: hibernate-core/src/main/java/org/hibernate/persister/entity/AbstractEntityPersister.java [hunk 2]
```diff
@@ -3926,20 +3927,26 @@ public Boolean isTransient(Object entity, SharedSessionContractImplementor sessi
...
+								.getLoadContexts()
+								.findLoadingEntityEntry( new EntityKey( id, this ) );
+						if ( loadingEntityEntry != null && loadingEntityEntry.getEntityInstance() == entity ) {
+							return false;
+						}
+					}
+					final Boolean unsaved = identifierMapping.getUnsavedStrategy().isUnsaved( id );
+					if ( unsaved != null && !unsaved ) {
+						throw new VersionValueException(
+								"A Persisted Versioned Entity cannot have an unsaved version value" );
```
Comment (gavinking) [inline hibernate-core/src/main/java/org/hibernate/persister/entity/AbstractEntityPersister.java:3946]: How about:

    "Transient entity with generated id '" + id + "' has an uninitialized value '" + version + "' for its version attribute"

File: hibernate-core/src/test/java/org/hibernate/orm/test/version/AssignedIdUpdateWithNullVersionTest.java [hunk 3]
```diff
@@ -0,0 +1,434 @@
...
+				AssignedIdUpdateWithNullVersionTest.TableGeneratedIdItem.class,
+				AssignedIdUpdateWithNullVersionTest.AssignedIdItem.class,
+				AssignedIdUpdateWithNullVersionTest.UUIDIdItem.class,
+		}
+)
+@SessionFactory
+@JiraKey("HHH-16586")
+@RequiresDialectFeature(feature = DialectFeatureChecks.SupportsSequences.class)
+@RequiresDialectFeature(feature = DialectFeatureChecks.SupportsIdentityColumns.class)
+public class AssignedIdUpdateWithNullVersionTest {
```
Comment (gavinking) [inline hibernate-core/src/test/java/org/hibernate/orm/test/version/AssignedIdUpdateWithNullVersionTest.java:41]: How about `TransientEntityWithNullVersionTest`.

**Review by gavinking — CHANGES_REQUESTED** [review]

**Comment by gavinking** [comment]
> Are we fixing something here that we shouldn't fix?

For the record I read to fast. The change proposed by @dreab8 is good.

Basically, this exception is only thrown when:

- the non-null generated id indicates that the entity is detached, but
- the version is uninitialized.

Which makes perfect sense.

File: hibernate-core/src/main/java/org/hibernate/persister/entity/AbstractEntityPersister.java (hunk 2 above)
Comment (gavinking) [reply hibernate-core/src/main/java/org/hibernate/persister/entity/AbstractEntityPersister.java:3946]: Perhaps even just:

> "Detached entity with generated id '" + id + "' has uninitialized version value '" + version + "'"

File: hibernate-core/src/main/java/org/hibernate/exception/VersionValueException.java (hunk 1 above)
Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/exception/VersionValueException.java:1]: I'm going to remove `VersionValueException` and use as you suggest a `PropertyValueException`

**Pull request merged**